src/models/StyleTags.js

```diff
@@ -0,0 +1,399 @@
+      }
+    }
+
+    return str
+  }
+
+  return {
+    styleTag: el,
+    getIds: getIdsFromMarkersFactory(markers),
+    names,
+    insertMarker,
+    insertRules,
+    css,
+    toHTML: wrapAsHtmlTag(css, names),
+    toElement: wrapAsElement(css, names),
+    clone: throwCloneTagErr,
+  }
+}
+
+const makeBrowserTag = (el: HTMLStyleElement): Tag<Text> => {
+  const markers = Object.create(null)
+  const names = []
+
+  const insertMarker = id => {
+    const prev = markers[id]
+    if (prev !== undefined) {
+      return prev
+    }
+
+    const marker = (markers[id] = document.createTextNode(makeTextMarker(id)))
+    el.appendChild(marker)
+    return marker
+  }
+
+  const insertRules = (id, cssRules) => {
+    insertMarker(id).appendData(cssRules.join(' '))
+  }
+
+  const css = () => {
+    let str = ''
+    // eslint-disable-next-line guard-for-in
+    for (const id in markers) {
+      str += markers[id].data
+    }
+    return str
+  }
+
+  return {
+    styleTag: el,
+    getIds: getIdsFromMarkersFactory(markers),
+    names,
+    insertMarker,
+    insertRules,
+    css,
+    toHTML: wrapAsHtmlTag(css, names),
+    toElement: wrapAsElement(css, names),
+    clone: throwCloneTagErr,
+  }
+}
+
+const makeServerTag = (): Tag<[string]> => {
+  const markers = Object.create(null)
+  const names = []
+
+  const insertMarker = id => {
+    const prev = markers[id]
+    if (prev !== undefined) {
+      return prev
+    }
+
+    return (markers[id] = [makeTextMarker(id)])
+  }
+
+  const insertRules = (id, cssRules) => {
+    const marker = insertMarker(id)
+    marker[0] += cssRules.join(' ')
+  }
+
+  const css = () => {
+    let str = ''
+    // eslint-disable-next-line guard-for-in
+    for (const id in markers) {
+      str += markers[id][0]
+    }
+    return str
+  }
+
+  const tag = {
+    styleTag: null,
+    getIds: getIdsFromMarkersFactory(markers),
+    names,
+    insertMarker,
+    insertRules,
+    css,
+    toHTML: wrapAsHtmlTag(css, names),
+    toElement: wrapAsElement(css, names),
+    clone() {
+      return {
+        ...tag,
+        names: [...names],
+        markers: { ...markers },
+      }
+    },
+  }
+
+  return tag
+}
+
+export const makeTag = (
+  target: ?HTMLElement,
+  lastEl: ?HTMLStyleElement,
+  forceServer?: boolean
+): Tag<any> => {
+  if (IS_BROWSER && !forceServer) {
+    const el = makeStyleTag(target, lastEl)
+    if (DISABLE_SPEEDY) {
+      return makeBrowserTag(el)
+    } else {
+      return makeSpeedyTag(el)
+    }
+  }
+
+  return makeServerTag()
+}
+
+/* TODO: Turn into fully functional composition (tag.names) */
+export const makeRehydrationTag = (
+  tag: Tag<any>,
+  els: HTMLStyleElement[],
+  extracted: ExtractedComp[],
+  names: string[],
+  immediateRehydration: boolean
+): Tag<any> => {
+  let isReady = false
+
+  /* rehydration function that adds all rules to the new tag */
+  const rehydrate = () => {
+    /* only rehydrate once */
+    if (isReady) {
+      return
+    } else {
```

Comment:
nit: don't need the else block

Reply:
Yep, it’s more of a stylistic choice but not in line obv